sqlglot/optimizer/qualify_columns.py

```diff
@@ -533,12 +533,21 @@ def _convert_columns_to_dots(scope: Scope, resolver: Resolver) -> None:
                 # The struct is already qualified, but we still need to change the AST
                 column_table = root
                 root, *parts = parts
+                was_qualified = True
             else:
                 column_table = resolver.get_table(root.name)
+                was_qualified = False
 
             if column_table:
                 converted = True
-                column.replace(exp.Dot.build([exp.column(root, table=column_table), *parts]))
+                new_column = exp.column(root, table=column_table)
+                if dot_parts := column.meta.get("dot_parts"):
+                    # Remove the actual column parts from the rest of dot parts
+                    new_column.meta["dot_parts"] = dot_parts[2 if was_qualified else 1 :]
```

Comment:
Can we have multi-part columns where `db` and `catalog` are legit and not parts of a dot sequence? Is this handled here?

Reply:
Ok, probably good as-is then.